src/GafferScene/PrimitiveVariableQuery.cpp

```diff
@@ -0,0 +1,553 @@
+
+		if( input == childQueryPlug->namePlug() )
+		{
+			addChildPlugsToAffectedOutputs( valuePlug, outputs );
+			outputs.push_back( outputPlug->getChild< const Gaffer::BoolPlug >( g_existsPlugIndex ) );
+			outputs.push_back( outputPlug->getChild< const Gaffer::StringPlug >( g_typePlugIndex ) );
+			outputs.push_back( outputPlug->getChild< const Gaffer::IntPlug >( g_interpolationPlugIndex ) );
+		}
+		else if( childQueryPlug->valuePlug() == input || childQueryPlug->valuePlug()->isAncestorOf( input ) )
+		{
+			outputs.push_back(
+				correspondingPlug(
+					static_cast< const Gaffer::ValuePlug* >( childQueryPlug->valuePlug< const Gaffer::ValuePlug >() ),
+					IECore::runTimeCast< const Gaffer::ValuePlug >( input ),
+					valuePlug
+				)
+			);
+		}
+	}
+}
+
+void PrimitiveVariableQuery::hash( const Gaffer::ValuePlug* const output, const Gaffer::Context* const context, IECore::MurmurHash& h ) const
+{
+	Gaffer::ComputeNode::hash( output, context, h );
+
+	if( output == internalObjectPlug() )
+	{
+		const std::string loc = locationPlug()->getValue();
+
+		if( ! loc.empty() )
+		{
+			const ScenePlug* const splug = scenePlug();
+			const ScenePlug::ScenePath path = ScenePlug::stringToPath( loc );
+
+			if( splug->exists( path ) )
+			{
+				h.append( splug->objectHash( path ) );
+			}
+		}
+	}
+	else if( outPlug()->isAncestorOf( output ) )
+	{
+		const Gaffer::ValuePlug* const outputPlug = outPlug( output );
+		const Gaffer::NameValuePlug* const childQueryPlug = queryPlug( outputPlug );
+
+		if(
+			( output == outputPlug->getChild( g_existsPlugIndex        ) ) ||
+			( output == outputPlug->getChild( g_typePlugIndex          ) ) ||
+			( output == outputPlug->getChild( g_interpolationPlugIndex ) ) )
+		{
+			internalObjectPlug()->hash( h );
+			childQueryPlug->namePlug()->hash( h );
+			return;
+		}
+
+		const Gaffer::ValuePlug* const valuePlug = outputPlug->getChild< const Gaffer::ValuePlug >( g_valuePlugIndex );
+		if( valuePlug && ( valuePlug->isAncestorOf( output ) || output == valuePlug ) )
+		{
+			internalObjectPlug()->hash( h );
+			childQueryPlug->namePlug()->hash( h );
+			correspondingPlug(
+				valuePlug,
+				output,
+				static_cast< const Gaffer::ValuePlug* >( childQueryPlug->valuePlug() )
+			)->hash( h );
+		}
+	}
+}
+
+void PrimitiveVariableQuery::compute( Gaffer::ValuePlug* const output, const Gaffer::Context* const context) const
+{
+	if( output == internalObjectPlug() )
+	{
+		// NOTE : cache the primitive at the specified location in the input scene
+		//        as all queries operate on this same primitive.
+
+		IECore::ConstObjectPtr object( IECore::NullObject::defaultNullObject() );
+		const std::string loc = locationPlug()->getValue();
+
+		if( ! loc.empty() )
+		{
+			const ScenePlug* const splug = scenePlug();
+			const ScenePlug::ScenePath path = ScenePlug::stringToPath( loc );
+
+			if( splug->exists( path ) )
+			{
+				object = IECore::runTimeCast< const IECoreScene::Primitive >( splug->object( path ) );
```

Comment:
This assigns `nullptr` to `object` when the cast fails, leading us to call `setValue( nullptr )`, which is forbidden. In my test here, it lead to a "Compute did not set plug value" exception, but it could lead anywhere. Would be good to have some test coverage for this - I'd suggest using Camera as a good example of a prevalent non-primitive object type. Rather than throw errors, queries on non-primitive types should just output equivalent results to primitive types where the variable is missing.

Reply:
done